File: java/Strings/Day4/SortCharByFreq.java
package Day4;
import java.util.*;

public class SortCharByFreq {
    public String frequencySort(String s) {
        HashMap<Character, Integer> freq = new HashMap<>();

        for (char ch : s.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }

        List<List<Character>> buckets = new ArrayList<>();
        for (int i = 0; i <= s.length(); i++) {
            buckets.add(new ArrayList<>());
        }


        for (Map.Entry<Character, Integer> entry : freq.entrySet()) {
            char ch = entry.getKey();
            int f = entry.getValue();
            buckets.get(f).add(ch);
        }



        StringBuilder ans = new StringBuilder();

        for (int f = s.length(); f >= 1; f--) {
            for (char ch : buckets.get(f)) {
                ans.append(String.valueOf(ch).repeat(f));
            }
        }

        return ans.toString();
    }



    

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter string: ");
        String s = sc.nextLine();

        SortCharByFreq obj = new SortCharByFreq();
        String result = obj.frequencySort(s);

        System.out.println("Sorted by frequency: " + result);




        sc.close();
    }
}

File: python/Strings/day4/sortCharByFreq.py
def frequency_sort(s):
    freq = {}
    for ch in s:
        freq[ch] = freq.get(ch, 0) + 1

    buckets = [[] for _ in range(len(s) + 1)]

    for ch, f in freq.items():
        buckets[f].append(ch)

    
    ans = []

    for f in range(len(s), 0, -1):
        for ch in buckets[f]:
            ans.append(ch * f)
    
    return "".join(ans)






# -------- main --------
if __name__ == "__main__":
    s = input("Enter string: ")
    ans = frequency_sort(s)
    print("Sorted by frequency:", ans)